### ISSUE TITLE
APIs stabilization metabug

### ISSUE THREAD

**Issue by alexcrichton** [rust-lang/rust#24028]
We will eventually need a process for API stabilization planning and discussion, but while we're first getting used to the 6 week cadence, we'll use this issue for tracking.

### Planned for 1.0 stable

* [x] `str::Utf8Error` (switch from enum to opaque struct, with accessor for "last valid byte position")
* [x] `io::Error::from_raw_os_error`
* [x] `Clone::clone_from` (likely)

### Planned for 1.1 beta

* [x] `str_words` -- see #15628

### Planned for 1.1 stable

* [ ] `thread::catch_panic`
* [ ] `time::Duration`
* [x] debug builders
* [ ] `collections::Bound`
* [ ] `One/Zero` + .sum / .product
* [ ] `iter::StepBy` -- decide what to do on a negative step
* [ ] `Box::from_raw`, `boxed::into_raw`
* [ ] `char::encode_utf{8,16}`

### APIs with known problems/requiring more thought

* [ ] `iter::order` -- would prefer to use comparison operators
* [ ] `alloc::heap` -- huge risks for future tracing support etc.
* [ ] `IteratorExt::{min,max}_by` --  #15311
* [ ] `<[_]>::position_elem`
* [ ] `#[bench]` -- want more pluggable framework
* [ ] `std::slice::bytes` -- wants more design work
* [ ] `char::width`
* [ ] `Read::chars` -- has problematic error type; should this even be provided given how slow it is?
* [ ] `iter::range_inclusive` -- replace with adapters
* [ ] `collections::{BitVec, VecMap}` -- need clearer policy on what ships in collections
* [ ] `mem::copy{,_mut}_lifetime`, `ptr::Unique` -- wants more general design for "safer unsafe patterns"

### Slated for eventual deprecation/removal (if unstable)

* [ ] `Vec::push_all`
* [ ] `Vec::from_raw_buf`
* [ ] `FromPrimitive`/`ToPrimitive`/`NumCast` (will be replaced with more general failable cast)

### Far-future

* [ ] `raw::TraitObject`
* [ ] `intrinsics::volatile_set_memory`
* [ ] `intrinsics::type_name`

**Comment by alexcrichton** [comment]
triage: P-high (1.1)

**Comment by steveklabnik** [comment]
`str_words` is very small, might be worthwhile :)

**Comment by bstrie** [comment]
`std::str::from_utf8` is marked stable, but it can return `std::str::Utf8Error` which is still unstable. Need to correct this oversight, as well as be sure to audit all stable functions to make sure their return types are stable.

**Comment by mitsuhiko** [comment]
`push_all` on vectors really needs to become stable.  I'm trying to replace my uses with that extend/into_iter or iter/cloned dance currently and it's not just slow, it's also terrible to look at.

**Comment by BurntSushi** [comment]
Some nice-to-haves:

* [`char::width`](http://doc.rust-lang.org/nightly/std/primitive.char.html#method.width)
* [`accessed` and `modified` on `fs::Metadata`](http://doc.rust-lang.org/nightly/std/fs/struct.Metadata.html) (I guess this is probably blocking on stabilizing some kind of `Duration` type.)

**Comment by kinghajj** [comment]
Any chance that `alloc::heap` can be made stable? The `deque` crate uses this to allocate its buffers.

**Comment by vhbit** [comment]
`convert` to pass `Path` to C code.

**Comment by sfackler** [comment]
I'd like to stabilize the debug builders.

**Comment by mrmonday** [comment]
I'm not sure that this is the right place to mention it, but http://doc.rust-lang.org/std/io/struct.Error.html#method.from_os_error is marked with "unclear whether this function is necessary" - I would argue it is necessary for the following reasons:

 * The standard library does not (and probably should not) provide abstractions for all OS APIs
 * Many libraries/miscellaneous kernel modules (not just OS APIs) return OS error codes
 * Third party libraries (i.e. not standard library) that wrap these APIs should be able to provide an API which is consistent with what the standard library offers (having to duplicate `std::io::Error` for each different library would be a shame).

Arguments against:
 * Users could mistakenly try to use it for any FFI error codes, not just OS error codes, leading to confusion. I'm not sure if/how this could be avoided.

**Comment by skade** [comment]
I had to do a change because [type_name](http://static.rust-lang.org/doc/master/std/intrinsics/fn.type_name.html) is exported into std, but not marked stable :/. It's not that I absolutely need it, but I have a usecase for something with it's guarantees: a stable identifier for a specific type between versions.

**Comment by pyfisch** [comment]
>  * [ ] `FromPrimitive` - e.g. casting from an integer to an enum
Great, that it will be added again. Hyper used it for both HTTP status codes and quality values, it was removed to make hyper run on beta. But Hyper also needs `ToPrimitive`, will this also stabilized, or is it planned to deprecate it?

**Comment by tomaka** [comment]
`Read::chars` is very useful when you want to lex/parse some UTF-8 input.

**Comment by mvdnes** [comment]
Something stabilizing `#![no_std]` and `extern crate core;` would be great.
For normal crates it isn't very useful, but it would be nice to compile my 'kernel' with a stable compiler.

**Comment by nagisa** [comment]
> accessed and modified on fs::Metadata (I guess this is probably blocking on stabilizing some kind of Duration type.)

`Duration` for an absolute time stamp sounds odd.

**Comment by maximih** [comment]
I also need `char::width`

**Comment by apasel422** [comment]
`std::collections::Bound` and its variants have no stability marker, but I assume this is waiting on changes or additions to the `Range` types.

**Comment by jedisct1** [comment]
`collections::BitVec` would be nice to stabilize.

**Comment by dnaq** [comment]
`std::intrinsics::volatile_set_memory`

**Comment by skade** [comment]
`test`, because `#[bench]` requires "test::Bencher" for its function signatures.

**Comment by csherratt** [comment]
`std::sync::Semaphore`

**Comment by SimonSapin** [comment]
> `str_words`

I think we should just remove this one. It’s unclear what "a word" is or should be. Unicode has an entire document ([UAX #29 Text Segmentation](http://www.unicode.org/reports/tr29/)) with a non-trivial algorithm to standardize this, but I think an implementation of it belongs more on crates.io than in `std`.

Instead, we could have "split a string on whitespace" in `std`, based on `char::is_whitespace`. This is much less ambiguous. It could be either a stand-alone method, or a custom `std::str::Pattern` implementation to be used with `str::split`.

**Comment by skade** [comment]
I agree with @SimonSapin there. I think a good crates.io library is what we need. https://github.com/rust-lang/rfcs/issues/797#issuecomment-73041246

**Comment by anonova** [comment]
`Vec::from_raw_buf`

**Comment by skade** [comment]
@anonova This is quite well expressed by `from_raw_parts` combined with `to_vec`

```
use std::slice::from_raw_parts;
from_raw_parts(value, length as usize).to_vec()
```

**Comment by reem** [comment]
@skade that requires a copy of all the contents, which is often not desirable.

**Comment by jedisct1** [comment]
I would really love to see `Vec::from_raw_buf()` to be stabilized as well.

**Comment by TyOverby** [comment]
I'm blocking on the whole `unicode` crate.

It also looks like NumCast and RawSlice are blocking for me as well.

**Comment by SimonSapin** [comment]
> I'm blocking on the whole `unicode` crate.

What features specifically? At some point I want to move them to crates.io

**Comment by carllerche** [comment]
I'm pretty keen on getting `TraitObject` or some API to work with them stable.

**Comment by Benjamin-L** [comment]
`libc` is sometimes required to do things although I doubt that it will be stable for a long time. For example, signal handling requires it. A better solution would be to put wrappers around `libc` functionality in the standard library.

**Comment by bluss** [comment]
> I had to do a change because type_name is exported into std, but not marked stable. It's not that I absolutely need it, but I have a usecase for something with it's guarantees: a stable identifier for a specific type between versions. I'd prefer using it.

@skade It sounds unlikely that Rust would guarantee that the type names were stable across versions, even if the function would be marked stable.

**Comment by skade** [comment]
@bluss I don't understand. How would stable public type names and user-provided type names change?

**Comment by bluss** [comment]
When exporting compiler internals right out I'd always like to use that caveat first, it's much easier to not promise anything. It seems the output is not stable now, given that the type name of "struct Foo" is `Foo` and of trait "Copy" it is `core::marker::Copy` which apart from being inconsistent with using path or not, leaks implementation details (location inside implementation crate libcore)

**Comment by ogham** [comment]
@TyOverby: As someone who's dependent on `str#width`, I'd really like a stable unicode crate. Is there anything blocking this function in particular?

**Comment by jgallagher** [comment]
I'd like to use `ptr::Unique` as in [this PR](https://github.com/jgallagher/rusqlite/pull/26/files) for holding on to the handle to a SQLite database.

**Comment by ajfrantz** [comment]
One vote for `core::str::StrExt` to be available out of the box.

**Comment by SimonSapin** [comment]
@ajfrantz, `core::str::StrExt` is replaced with inherent methods on the `str` type: http://doc.rust-lang.org/std/primitive.str.html

**Comment by ajfrantz** [comment]
@SimonSapin Ah, perfect.  If I just search 'contains' in the API docs it doesn't make that very obvious.

Thanks!

**Comment by SimonSapin** [comment]
Yes, that’s a bug: https://github.com/rust-lang/rust/issues/23511

**Comment by blackbeam** [comment]
Vote for `char::encode_utf(8|16)`.

**Comment by nagisa** [comment]
> Vote for char::encode_utf(8|16).

There’s the `encoding` crate.

**Comment by blackbeam** [comment]
> There’s the `encoding` crate.

Thanks!

**Comment by SimonSapin** [comment]
> Vote for char::encode_utf(8|16).
As a work-around, I’ve replaced `encode_utf8` with:

```rust
        let mut utf_8 = [0u8; 4];
        let bytes_written = {
            let mut buffer = &mut utf_8[..];
            write!(buffer, "{}", c).unwrap();
            4 - buffer.len()
        };
        self.write_str(unsafe { mem::transmute(&utf_8[..bytes_written]) })
```

As to, `encode_utf16`, you could just inline it:

```rust
if ch < '\u{10000}' {
    &[ch as u16][..]
} else {
    ch -= 0x1_0000;
    &[0xD800 | ((ch >> 10) as u16),
      0xDC00 | ((ch as u16) & 0x3FF)][..]
}
```

**Comment by blackbeam** [comment]
Thanks, Simon!

Also vote for `TcpStream::set_keepalive`.

**Comment by xitep** [comment]
`test`

**Comment by daboross** [comment]
@vhbit It's totally possible at the moment to get from `OsStr` to `CString` on rust stable, it's just a bit indirect I guess, though it shouldn't really be any more costly performance wise than the unstable APIs.

```
use std::ffi::CString;
use std::path::Path;

fn main() {
    let path = Path::new(".");
    let os_str = path.as_os_str();
    let str = os_str.to_str().unwrap(); // fails if invalid unicode
    let c_string = CString::new(str).unwrap(); // fails if null bytes
}
```

**Comment by SimonSapin** [comment]
`std::os::unix::ffi::OsStrExt::as_bytes` is stable and does not require UTF-8 well-formedness, but only exists on Unix. (I’m not sure what encoding a `CString` is supposed to be in on Windows anyway.)

**Comment by carllerche** [comment]
Another API that is important: `NonZero`. Without it, structs that used to be 1ptr in size have grown to 4x the size (because of nesting).

**Comment by daboross** [comment]
Personally, `IntoCow` (or `Into<Cow>`) being stabilized would be awesome, it's something that would be really useful when creating library APIs.

**Comment by BurntSushi** [comment]
@daboross `Into<Cow>` is stable: http://doc.rust-lang.org/std/convert/trait.Into.html and http://doc.rust-lang.org/std/borrow/enum.Cow.html

**Comment by daboross** [comment]
@BurntSushi Oh, I guess they are. For some reason I assumed that there would have to be some custom implementations of Into<Cow> like there were listed on IntoCow<> for it to work, but Into<Cow<'static, str>> seems to work fine. Thanks!

**Comment by TyOverby** [comment]
@SimonSapin In `unicode`, I'm just using the `str::utf8_char_witdth` function.

**Comment by dirvine** [comment]
I would like to add a request for ``` std::sync::Weak``` if at all possible.

**Comment by carllerche** [comment]
I need `Error::from_os_error`. This allows me to convert from a nix error -> `io::Error`

**Comment by cgaebel** [comment]
The following need to be stabilized before I can ship `iobuf` working on rustc beta:

`test::{black_box, Bencher}`
`alloc::heap::allocate`
`alloc::heap::deallocate`
`core::nonzero`
`#![feature(unsafe_no_drop_flag)]` needs to either be stabilized, or removed. There will be a perf cost if removed before non-filling drop lands.

**Comment by bfops** [comment]
I get the impression the `entry` API was never really stabilized? `entry` should take a reference to a key, and only did a (potentially expensive) `clone` when `VacantEntry::insert` is called.

**Comment by daboross** [comment]
@carllerche Does `io::Error::last_os_error()` not do what is needed? It literally just calls `from_os_error` with the last error. I'm not experienced with working with system APIs at all, but I *think* that might be usable for a situation where you have a nix error.

**Comment by carllerche** [comment]
@daboross it does not. By the time I need it, There is no guarantee that the last error is the right one.

**Comment by daboross** [comment]
Ah, ok - just wondering, thanks!

**Comment by wackywendell** [comment]
I have now run into needing something like `stepBy` in two different crates. Is there a stable way of doing `(1..10).step_by` that isn't too annoying?

**Comment by ruud-v-a** [comment]
A [program](https://github.com/ruud-v-a/robigo-luculenta) of mine is blocked on `AdditiveIterator`. A [library](https://github.com/ruud-v-a/claxon) of mine is blocked on `AdditiveIterator` and `From`/`ToPrimitive`. `AdditiveIterator` can easily be worked around, but it would be nice to have.

**Comment by aturon** [comment]
It's going to take a few release cycles before we fully hit our stride and have a good process for stabilization planning. For the time being, I've cleaned up this issue after discussion with @alexcrichton. This should now give you a clearer picture for the immediate plans.

**Comment by drewcrawford** [comment]
Is there any way to get separate issues to track these?  I want to subscribe to like 3 of them but I don't want to pile on with everybody who wants every API stabilized in one issue.

**Comment by SimonSapin** [comment]
> str_words -- see #15628 

Please see https://github.com/rust-lang/rfcs/pull/1054

**Comment by bstrie** [comment]
`std::rc::Weak`. The reason given for instability, "weak pointers may not belong in this module", seems easy enough to address. An example of a project blocked on this: https://github.com/aepsil0n/carboxyl/issues/47 (from http://blog.ebopp.de/blog/2015/04/23/interactive-2d-apps/).

**Comment by GBGamer** [comment]
Vote for collections::VecMap, it's really useful one of my projects.

**Comment by TyOverby** [comment]
Another vote for VecMap.

**Comment by nwin** [comment]
It is a while since I looked last time into this issue, but is the current plan really to remove `push_call(slice)`. Only to replace it with `extend(slice.iter().cloned())` or  `extend(slice.iter().map(|&v| v))`? That looks overly verbose and ambiguous.

**Comment by bluss** [comment]
There's an RFC ot make `.extend(slice)` work (while still keeping the present extend impls too).

**Comment by bluss** [comment]
I'd propose these as APIs with problems:

- slice::permutations, Permutations, ElementSwaps
- slice::next_permutation, slice::prev_permutation

They have unclear performance impact and utility. They candidates for moving to a third party crate.

As noted on the forum thread, my wishlist for (not far) future stabilization:

- intrinsics::unreachable

**Comment by cybergeek94** [comment]
I can't build [`img_hash`][img_hash] on Stable without `BitVec` unless it gets externalized.

**Comment by cybergeek94** [comment]
Also need `UnsafeCell::value`

**Comment by bluss** [comment]
@cybergeek94 bitvec is now on crates.io, crate bit-vec

**Comment by TyOverby** [comment]
I'd also really like `net::lookup_host`.

**Comment by reem** [comment]
@bluss and others, the `unreachable` crates.io crate now provides a works-on-stable version of `intrinsics::unreachable` based on the `void` crate and should have a similar effect on code generation.

**Comment by GBGamer** [comment]
Could we add `core::nonzero::NonZero` to the list, to add to `std` and stabilize? It's super useful, and I can't think of any reason to not stabilize it. It's not like it will change.

**Comment by bluss** [comment]
@alexcrichton  @reem The adjustments to new llvm in issue #26025 might affect the codegen for `match Void { /* empty */ }`, it least changes the IR emitted.

**Comment by eddyb** [comment]
@GBGamer `NonZero` doesn't have a serious API at the moment, there's no way the current hacky state would be stabilized.

**Comment by GBGamer** [comment]
@eddyb It would be nice to at least re-export from std, if possible, even if it is unstable. Having to `extern crate core` is just an annoyance that could easily be fixed.

Also, what kind of API could we design around NonZero? Serious question, I can't think of anything besides `new(x: T) //-> panics if x is zero` and `new_unchecked(x: T) // UB if x is zero`.

**Comment by eddyb** [comment]
@GBGamer first off, the `Deref` impl is just broken, it doesn't really fit attaching metadata to loads the way it's supposed to be. A `get` method (like `Cell` has) is much better suited.

And then, there's the following constructor:
```rust
impl<T: Zeroable> NonZero<T> {
    fn new(x: T) -> Option<NonZero<T>> {
        if x.is_zero() {
            None
        } else {
            Some(NonZero(x))
        }
    }
}
```
Completely safe, should compile down to nothing, and the panic behaviour you describe could be achieved by just calling `.unwrap()` on the result.

It could be used like this:
```rust
extern {
    fn malloc(NonZero<usize>) -> *mut u8;
}

fn oom() -> ! { print_oom_message(); abort() }

const EMPTY: *mut u8 = 1 as *mut u8;
fn alloc(size: usize) -> NonZero<*mut u8> {
    NonZero::new(size).map_or(EMPTY, |size| {
        NonZero::new(malloc(size)).or_else(oom)
    })
}
```

**Comment by GBGamer** [comment]
@eddyb alright, makes total sense. I would like to start working on that
soon, then, perhaps after finals.

On Tue, Jun 16, 2015 at 11:26 PM, Eduard Burtescu <notifications@github.com>
wrote:

> @GBGamer <https://github.com/GBGamer> first off, the Deref impl is just
> broken, it doesn't really fit attaching metadata to loads the way it's
> supposed to be. A get method (like Cell has) is much better suited.
>
> And then, there's the following constructor:
>
> impl<T: Zeroable> NonZero<T> {
>     fn new(x: T) -> Option<NonZero<T>> {
>         if x.is_zero() {
>             None
>         } else {
>             Some(NonZero(x))
>         }
>     }
> }
>
> Completely safe, should compile down to nothing, and the panic behaviour
> you describe could be achieved by just calling .unwrap() on the result.
>
> It could be used like this:
>
> extern {
>     fn malloc(NonZero<usize>) -> *mut u8;
> }
> fn oom() -> ! { print_oom_message(); abort() }
> const EMPTY: *mut u8 = 1 as *mut u8;fn alloc(size: usize) -> NonZero<*mut u8> {
>     NonZero::new(size).map_or(EMPTY, |size| {
>         NonZero::new(malloc(size)).or_else(oom)
>     })
> }
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rust-lang/rust/issues/24028#issuecomment-112670509>.
>



-- 
"Unjust laws exist; shall we be content to obey them, or shall we endeavor
to amend them, and obey them until we have succeeded, or shall we
transgress them at once?"
        -- Henry David Thoreau

**Comment by cybergeek94** [comment]
Is there anything blocking the new static methods on `Rc` from being stabilized? There's no notes for them.

**Comment by cybergeek94** [comment]
Same with `slice::position_elem`.

**Comment by bluss** [comment]
@cybergeek94 For `position_elem` I just guess that it's related to the want for a more extended searching API for slices. For string slices we have patterns, and you can `.find()` a char, substring, set of chars etc. The question is what we want to do with `&[_]` and stabilizing `position_elem` is affected by that.

**Comment by mitsuhiko** [comment]
Not sure if this is the right place to bring it up, but `catch_panic` seems a bit misplaced in the thread module given that it does not actually have anything directly to do with threads.  Would it not make sense to put this somewhere else?  I thought from reading the docs that it spawns a thread and captures the panic.

**Comment by sfackler** [comment]
That's what it used to do! But yeah, it should probably be moved (maybe to `ffi`?)

**Comment by aturon** [comment]
Discussion of `catch_panic` is [here](https://github.com/rust-lang/rfcs/pull/1236).

**Comment by aturon** [comment]
I'm going to close this issue out in favor of [per-feature tracking issues](https://internals.rust-lang.org/t/psa-stabilization-tracking-on-the-libs-team/2493), which have now been created for the bulk of unstable std APIs. Please comment on the individual issues with your use cases, stabilization desire, and feedback!
